File: swift-bridge/Sources/CoreAnimationBridge/Renderer.swift
import CoreGraphics
import CoreVideo
import Foundation
import Metal
import QuartzCore

@_cdecl("ca_renderer_new")
public func ca_renderer_new(_ textureHandle: UnsafeMutableRawPointer?, _ queueHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let texture: MTLTexture = caBorrow(textureHandle) else { return nil }
    let queue: MTLCommandQueue? = caBorrow(queueHandle)
    let options: [AnyHashable: Any]? = queue.map { [kCARendererMetalCommandQueue: $0] }
    return caRetain(CARenderer(mtlTexture: texture, options: options))
}

@_cdecl("ca_renderer_set_layer")
public func ca_renderer_set_layer(_ handle: UnsafeMutableRawPointer?, _ layerHandle: UnsafeMutableRawPointer?) {
    guard let renderer: CARenderer = caBorrow(handle) else { return }
    let layer: CALayer? = caBorrow(layerHandle)
    renderer.layer = layer
}

@_cdecl("ca_renderer_get_bounds")
public func ca_renderer_get_bounds(_ handle: UnsafeMutableRawPointer?, _ outRect: UnsafeMutableRawPointer?) -> Bool {
    guard let renderer: CARenderer = caBorrow(handle) else { return false }
    return caWriteRect(renderer.bounds, out: outRect)
}

@_cdecl("ca_renderer_set_bounds")
public func ca_renderer_set_bounds(_ handle: UnsafeMutableRawPointer?, _ x: Double, _ y: Double, _ width: Double, _ height: Double) {
    guard let renderer: CARenderer = caBorrow(handle) else { return }
    renderer.bounds = CGRect(x: x, y: y, width: width, height: height)
}

@_cdecl("ca_renderer_begin_frame")
public func ca_renderer_begin_frame(_ handle: UnsafeMutableRawPointer?, _ time: Double, _ timestamp: UnsafeMutableRawPointer?) {
    guard let renderer: CARenderer = caBorrow(handle) else { return }
    if let timestamp {
        renderer.beginFrame(atTime: time, timeStamp: timestamp.assumingMemoryBound(to: CVTimeStamp.self))
    } else {
        renderer.beginFrame(atTime: time, timeStamp: nil)
    }
}

@_cdecl("ca_renderer_update_bounds")
public func ca_renderer_update_bounds(_ handle: UnsafeMutableRawPointer?, _ outRect: UnsafeMutableRawPointer?) -> Bool {
    guard let renderer: CARenderer = caBorrow(handle) else { return false }
    return caWriteRect(renderer.updateBounds(), out: outRect)
}

@_cdecl("ca_renderer_render")
public func ca_renderer_render(_ handle: UnsafeMutableRawPointer?) {
    guard let renderer: CARenderer = caBorrow(handle) else { return }
    renderer.render()
}

@_cdecl("ca_renderer_end_frame")
public func ca_renderer_end_frame(_ handle: UnsafeMutableRawPointer?) {
    guard let renderer: CARenderer = caBorrow(handle) else { return }
    renderer.endFrame()
}

@_cdecl("ca_renderer_next_frame_time")
public func ca_renderer_next_frame_time(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let renderer: CARenderer = caBorrow(handle) else { return .infinity }
    return renderer.nextFrameTime()
}

@_cdecl("ca_renderer_set_destination")
public func ca_renderer_set_destination(_ handle: UnsafeMutableRawPointer?, _ textureHandle: UnsafeMutableRawPointer?) {
    guard let renderer: CARenderer = caBorrow(handle), let texture: MTLTexture = caBorrow(textureHandle) else { return }
    renderer.setDestination(texture)
}

@_cdecl("ca_renderer_render_at_time")
public func ca_renderer_render_at_time(_ handle: UnsafeMutableRawPointer?, _ time: Double) {
    guard let renderer: CARenderer = caBorrow(handle) else { return }
    CATransaction.flush()
    renderer.beginFrame(atTime: time, timeStamp: nil)
    renderer.render()
    renderer.endFrame()
}

@_cdecl("ca_texture_copy_bytes")
public func ca_texture_copy_bytes(_ textureHandle: UnsafeMutableRawPointer?, _ outBytes: UnsafeMutableRawPointer?, _ bytesPerRow: Int) -> Bool {
    guard let texture: MTLTexture = caBorrow(textureHandle), let outBytes else { return false }
    let region = MTLRegionMake2D(0, 0, texture.width, texture.height)
    texture.getBytes(outBytes, bytesPerRow: bytesPerRow, from: region, mipmapLevel: 0)
    return true
}

File: src/renderer.rs
use apple_cf::cg::{CGColorSpace, CGRect};
use apple_metal::{pixel_format, CommandQueue, MetalTexture};

use crate::display_link::CVTimeStamp;
use crate::error::CoreAnimationError;
use crate::layer::LayerLike;
use crate::private::handle_type;

handle_type!(Renderer);

impl Renderer {
    #[must_use]
    /// Creates a `CARenderer` targeting a Metal texture.
    pub fn new(texture: &MetalTexture, queue: Option<&CommandQueue>) -> Option<Self> {
        Self::new_with_color_space(texture, queue, None)
    }

    #[must_use]
    /// Creates a `CARenderer` targeting a Metal texture with an optional color space.
    pub fn new_with_color_space(
        texture: &MetalTexture,
        queue: Option<&CommandQueue>,
        color_space: Option<&CGColorSpace>,
    ) -> Option<Self> {
        unsafe {
            Self::from_raw(crate::ffi::ca_renderer_new_with_options(
                texture.as_ptr(),
                queue.map_or(core::ptr::null_mut(), CommandQueue::as_ptr),
                color_space.map_or(core::ptr::null_mut(), CGColorSpace::as_ptr),
            ))
        }
    }

    /// Sets the root layer rendered by the renderer.
    pub fn set_layer<L: LayerLike>(&self, layer: Option<&L>) {
        unsafe {
            crate::ffi::ca_renderer_set_layer(
                self.as_ptr(),
                layer.map_or(core::ptr::null_mut(), LayerLike::as_layer_ptr),
            )
        };
    }

    #[must_use]
    /// Returns the renderer's bounds.
    pub fn bounds(&self) -> CGRect {
        let mut rect = CGRect::zero();
        let ok = unsafe {
            crate::ffi::ca_renderer_get_bounds(
                self.as_ptr(),
                (&mut rect as *mut CGRect).cast::<core::ffi::c_void>(),
            )
        };
        if ok {
            rect
        } else {
            CGRect::zero()
        }
    }

    /// Sets the renderer's bounds.
    pub fn set_bounds(&self, rect: CGRect) {
        unsafe {
            crate::ffi::ca_renderer_set_bounds(
                self.as_ptr(),
                rect.origin.x,
                rect.origin.y,
                rect.size.width,
                rect.size.height,
            )
        };
    }

    /// Begins a renderer frame at the supplied media time.
    pub fn begin_frame(&self, time: f64, time_stamp: Option<&CVTimeStamp>) {
        unsafe {
            crate::ffi::ca_renderer_begin_frame(
                self.as_ptr(),
                time,
                time_stamp.map_or(core::ptr::null_mut(), |time_stamp| {
                    (time_stamp as *const CVTimeStamp)
                        .cast_mut()
                        .cast::<core::ffi::c_void>()
                }),
            )
        };
    }

    pub fn add_update_rect(&self, rect: CGRect) {
        unsafe {
            crate::ffi::ca_renderer_add_update_rect(
                self.as_ptr(),
                rect.origin.x,
                rect.origin.y,
                rect.size.width,
                rect.size.height,
            )
        };
    }

    #[must_use]
    /// Returns the bounds updated by the current renderer frame.
    pub fn update_bounds(&self) -> CGRect {
        let mut rect = CGRect::zero();
        let ok = unsafe {
            crate::ffi::ca_renderer_update_bounds(
                self.as_ptr(),
                (&mut rect as *mut CGRect).cast::<core::ffi::c_void>(),
            )
        };
        if ok {
            rect
        } else {
            CGRect::zero()
        }
    }

    /// Renders the current layer tree into the destination texture.
    pub fn render(&self) {
        unsafe { crate::ffi::ca_renderer_render(self.as_ptr()) };
    }

    /// Ends the current renderer frame.
    pub fn end_frame(&self) {
        unsafe { crate::ffi::ca_renderer_end_frame(self.as_ptr()) };
    }

    #[must_use]
    /// Returns the renderer-reported time for the next frame.
    pub fn next_frame_time(&self) -> f64 {
        unsafe { crate::ffi::ca_renderer_next_frame_time(self.as_ptr()) }
    }

    /// Sets the renderer destination texture.
    pub fn set_destination(&self, texture: &MetalTexture) {
        unsafe { crate::ffi::ca_renderer_set_destination(self.as_ptr(), texture.as_ptr()) };
    }

    /// Renders the current layer tree at the supplied media time.
    pub fn render_at_time(&self, time: f64) {
        unsafe { crate::ffi::ca_renderer_render_at_time(self.as_ptr(), time) };
    }
}

/// Copies the texel bytes of a CPU-readable 2D Metal texture, `width * bytes_per_pixel` per row.
pub fn read_texture_bytes(texture: &MetalTexture) -> Result<Vec<u8>, CoreAnimationError> {
    let format = texture.pixel_format();
    let bytes_per_pixel = match format {
        pixel_format::DEPTH16UNORM | pixel_format::DEPTH32FLOAT | pixel_format::STENCIL8 => None,
        format => apple_metal::bytes_per_pixel(format),
    }
    .ok_or_else(|| {
        CoreAnimationError::new(format!(
            "pixel format {format} has no CPU byte layout (compressed, depth, stencil or unknown)"
        ))
    })?;
    let bytes_per_row = texture
        .width()
        .checked_mul(bytes_per_pixel)
        .ok_or_else(|| CoreAnimationError::new("texture bytes_per_row overflowed"))?;
    let byte_count = bytes_per_row
        .checked_mul(texture.height())
        .filter(|count| isize::try_from(*count).is_ok())
        .ok_or_else(|| CoreAnimationError::new("texture byte count overflowed"))?;
    let mut bytes = vec![0_u8; byte_count];

    let status = unsafe {
        crate::ffi::ca_texture_copy_bytes(
            texture.as_ptr(),
            bytes.as_mut_ptr().cast::<core::ffi::c_void>(),
            bytes.len(),
            bytes_per_row,
            bytes_per_pixel,
        )
    };

    match status {
        0 => Ok(bytes),
        2 => Err(CoreAnimationError::new(format!(
            "texture type {} is not a plain 2D texture",
            texture.texture_type()
        ))),
        3 => Err(CoreAnimationError::new(format!(
            "texture storage mode {} is not CPU-accessible; blit it to a shared texture first",
            texture.storage_mode()
        ))),
        4 => Err(CoreAnimationError::new(
            "framebuffer-only textures cannot be read; disable framebuffer_only on the layer",
        )),
        _ => Err(CoreAnimationError::new(
            "failed to copy bytes from Metal texture",
        )),
    }
}
